core/src/main/java/com/linecorp/armeria/server/AggregatingDecodedHttpRequest.java

```diff
@@ -0,0 +1,174 @@
+
+    private final EventLoop eventLoop;
+    private final int id;
+    private final int streamId;
+    private final boolean keepAlive;
+    private final long maxRequestLength;
+    private final RequestHeaders headers;
+    private final RoutingContext routingCtx;
+    private final Routed<ServiceConfig> routed;
+    @Nullable
+    private ServiceRequestContext ctx;
+    @Nullable
+    private HttpHeaders trailers;
+    private long transferredBytes;
+
+    @Nullable
+    private HttpResponse response;
+    private boolean isResponseAborted;
+
+    AggregatingDecodedHttpRequest(EventLoop eventLoop, int id, int streamId, RequestHeaders headers,
+                                  boolean keepAlive, long maxRequestLength,
+                                  RoutingContext routingCtx,
+                                  Routed<ServiceConfig> routed) {
+        super(4);
+        this.headers = headers;
+        this.eventLoop = eventLoop;
+        this.id = id;
+        this.streamId = streamId;
+        this.keepAlive = keepAlive;
+        this.maxRequestLength = maxRequestLength;
+        this.routingCtx = routingCtx;
+        this.routed = routed;
+    }
+
+    @Override
+    public void init(ServiceRequestContext ctx) {
+        this.ctx = ctx;
+        ctx.logBuilder().increaseRequestLength(transferredBytes);
+        if (trailers != null) {
+            ctx.logBuilder().requestTrailers(trailers);
+        }
+    }
+
+    @Override
+    public RoutingContext routingContext() {
+        return routingCtx;
+    }
+
+    @Nonnull
+    @Override
+    public Routed<ServiceConfig> route() {
+        return routed;
+    }
+
+    @Override
+    public int id() {
+        return id;
+    }
+
+    @Override
+    public int streamId() {
+        return streamId;
+    }
+
+    @Override
+    public boolean isKeepAlive() {
+        return keepAlive;
+    }
+
+    @Override
+    public long maxRequestLength() {
+        return ctx != null ? ctx.maxRequestLength() : maxRequestLength;
+    }
+
+    @Override
+    public long transferredBytes() {
+        return transferredBytes;
+    }
+
+    @Override
+    public void increaseTransferredBytes(long delta) {
+        if (transferredBytes > Long.MAX_VALUE - delta) {
+            transferredBytes = Long.MAX_VALUE;
+        } else {
+            transferredBytes += delta;
+        }
+    }
+
+    @Override
+    public EventLoop defaultSubscriberExecutor() {
+        return eventLoop;
+    }
+
+    @Override
+    public boolean tryWrite(HttpObject obj) {
+        final boolean published = super.tryWrite(obj);
+
+        if (obj instanceof HttpData) {
+            ((HttpData) obj).touch(routingCtx);
+            if (obj.isEndOfStream()) {
+                close();
+            }
+        }
+        if (obj instanceof HttpHeaders) {
+            trailers = (HttpHeaders) obj;
+            close();
+        }
+        return published;
+    }
+
+    @Override
+    public void setResponse(HttpResponse response) {
```

Comment:
Because of class hierarchy, we can move this method to the super class.
Actually, I didn't want to create a delegating `DecodedHttpRequest` to avoid additional allocations.
But the advantage of avoiding allocation could be almost zero. It might be better to use a delegator than duplicate code.

Reply:
Ah, then it's fine as it is. 😉